src/hotspot/share/code/nmethod.cpp

```diff
@@ -1637,79 +1425,51 @@ oop nmethod::oop_at_phantom(int index) const {
         oop call_site = deps.argument_oop(0);
-        if (delete_immediately) {
-          assert_locked_or_safepoint(CodeCache_lock);
-          MethodHandles::remove_dependent_nmethod(call_site, this);
-        } else {
-          MethodHandles::clean_dependency_context(call_site);
-        }
+        MethodHandles::clean_dependency_context(call_site);
       } else {
         Klass* klass = deps.context_type();
         if (klass == NULL) {
           continue;  // ignore things like evol_method
         }
-        // During GC delete_immediately is false, and liveness
-        // of dependee determines class that needs to be updated.
-        if (delete_immediately) {
-          assert_locked_or_safepoint(CodeCache_lock);
-          InstanceKlass::cast(klass)->remove_dependent_nmethod(this);
-        } else if (klass->is_loader_alive()) {
-          // The GC may clean dependency contexts concurrently and in parallel.
-          InstanceKlass::cast(klass)->clean_dependency_context();
-        }
+        // During GC liveness of dependee determines class that needs to be updated.
+        // The GC may clean dependency contexts concurrently and in parallel.
+        InstanceKlass::cast(klass)->clean_dependency_context();
       }
     }
   }
 }
 
+void nmethod::post_compiled_method(CompileTask* task) {
+  task->mark_success();
+  task->set_nm_content_size(content_size());
+  task->set_nm_insts_size(insts_size());
+  task->set_nm_total_size(total_size());
+
+  // JVMTI -- compiled method notification (must be done outside lock)
+  post_compiled_method_load_event();
+
+  if (CompilationLog::log() != NULL) {
+    CompilationLog::log()->log_nmethod(JavaThread::current(), this);
+  }
+
+  const DirectiveSet* directive = task->directive();
+  maybe_print_nmethod(directive);
+}
+
 // ------------------------------------------------------------------
 // post_compiled_method_load_event
 // new method for install_code() path
 // Transfer information from compilation to jvmti
 void nmethod::post_compiled_method_load_event(JvmtiThreadState* state) {
-
-  // Don't post this nmethod load event if it is already dying
-  // because the sweeper might already be deleting this nmethod.
-  {
-    MutexLocker ml(CompiledMethod_lock, Mutex::_no_safepoint_check_flag);
-    // When the nmethod is acquired from the CodeCache iterator, it can racingly become zombie
-    // before this code is called. Filter them out here under the CompiledMethod_lock.
-    if (!is_alive()) {
-      return;
-    }
-    // As for is_alive() nmethods, we also don't want them to racingly become zombie once we
-    // release this lock, so we check that this is not going to be the case.
-    if (is_not_entrant() && can_convert_to_zombie()) {
-      return;
-    }
-    // Ensure the sweeper can't collect this nmethod until it become "active" with JvmtiThreadState::nmethods_do.
-    mark_as_seen_on_stack();
-  }
-
```

Comment:
Thank you! It's nice to see this code which still may have bugs removed.

Reply:
I'm glad you like it!